lib/queuePopulator/QueuePopulatorExtension.js

```diff
@@ -0,0 +1,100 @@
+const assert = require('assert');
+
+class QueuePopulatorExtension {
+    /**
+     * @constructor
+     * @param {Object} params - constructor params
+     * @param {Object} params.config - extension-specific
+     *   configuration object
+     * @param {Logger} params.logger - logger object
+     */
+    constructor(params) {
+        this.log = params.logger;
+        this._batch = null;
+    }
+
+    /**
+     * Subclasses should implement this function if asynchronous
+     * initial setup is required at startup
+     *
+     * @param {function} done - callback function
+     * @return {undefined}
+     */
+    setup(done) {
+        return process.nextTick(done);
+    }
+
+    /**
+     * Subclasses should implement this function if asynchronous
+     * code has to be executed at shutdown
+     *
+     * @param {function} done - callback function
+     * @return {undefined}
+     */
+    shutdown(done) {
+        return process.nextTick(done);
+    }
+
+    /**
+     * Prepare a kafka entry for publishing
+     *
+     * This function is meant to be called synchronously by subclasses
+     * in filter() implementation, called when new metadata log
+     * entries are received. It's extension's responsibility to decide
+     * which topic(s) to publish to and what payload to include in
+     * kafka messages then.
+     *
+     * @param {string} topic - topic name
+     * @param {string} key - key of the kafka entry, for keyed partitioning
+     * @param {string} message - kafka message
+     * @return {undefined}
+     */
+    publish(topic, key, message) {
+        assert(this._batch,
+               'logic error: QueuePopulatorExtension.publish() called ' +
+               'without an active batch. Please make sure it\'s called ' +
+               'synchronously from the filter() method.');
+
+        const kafkaEntry = { key: encodeURIComponent(key), message };
+        this.log.trace('queueing kafka entry to topic',
+                       { key: kafkaEntry.key, topic });
+        if (this._batch[topic] === undefined) {
+            this._batch[topic] = [kafkaEntry];
+        } else {
+            this._batch[topic].push(kafkaEntry);
+        }
+    }
+
+    /**
+     * Process a metadata log entry in extension subclass
+     *
+     * This method must be implemented by subclasses of QueuePopulatorExtension
+     * @param {Object} entry - metadata log entry
+     * @param {string} entry.bucket - bucket name
+     * @param {string} entry.type - entry type ('put'|'del')
+     * @param {string} entry.key - object key in log
+     * @param {string} entry.value - object value in log
+     * @return {undefined}
+     */
+    filter() {
+        assert(false,
```

Comment:
Maybe add a brief comment to note that this assertion is meant to ensure the extension has its own `filter` method.

Reply:
Since it's already explicit in the assert error message, and it's also in the JSDoc above, I think it would be double redundant :)